### ISSUE TITLE
Suggestion: standardize against prefixing rule names with "no-"

### ISSUE THREAD

**Issue by JoshuaKGoldberg** [typescript-eslint/typescript-eslint#203]
Many ESLint and TSLint rules start with "no-" to indicate a preference against some behavior. Some behaviors are always bad (e.g. `no-eval`), but others eventually are requested to have a preference towards. Having "no-" starting the rule name then makes it weird to add an option to encourage the behavior. For example:

* #103: [`no-parameter-properties`] Support enforcing the inverse
* #142: [`no-type-alias`]: Support enforcing the inverse

Having only some rules start with `no-` also makes them a little less discoverable. There are a few more characters to scan through, and alphabetical lists make a little less sense.

Since it's already a conversion to go from TSLint to ESLint, can there be a preference set for not starting rule names with "no-"? In the two cases above, it'd be better if they were `parameter-properties` and `type-alias`.

**Comment by armano2** [comment]
`no-` prefix in eslint is used for rules which `disallow` / `disable` some functionality.

Changing names of rules is considered breaking change and can be done only in major release

i don't disagree that some of rules should not be prefixed with `no-` (if options are going to be added for them)

https://eslint.org/docs/rules/

**Comment by JoshuaKGoldberg** [comment]
> Changing names of rules is considered breaking change

How about rule aliases? As in, having a `no-parameter-properties` file that just re-exports `parameter-properties`? At the next major version it could log a deprecation warning, and a major version after that, log an error, and one major version later, stop existing.

Just to clarify: for rules that aren't yet ported, would removing the `no-` be acceptable?

**Comment by bradzacher** [comment]
Yeah we need to knuckle out some guidance around this. Most of our `no-` rules were built because someone had an opinion about a ts feature and nobody had an opinion about ever wanting that feature on!

Now that we are much bigger, we should definitely nut out some guidance docs around this. `no-` rules only work in rare cases where it's bad to use a feature, or it makes no sense to have an inverse check (as in no-eval or no-explicit-any).

We can deprecate and redirect users to new rules to replace some of the `no-` rules without doing a major bump, but removing them obviously is breaking.

**Comment by Jessidhia** [comment]
Note also that you can't _just_ rename the rule from `no-` to "no `no-`". You also have to flip what `"always"` / `"never"` / et al mean.

Not saying it wouldn't be a good change, it's just not simple to do, and can't really be done with aliases unless there is some refactoring done.

**Comment by mysticatea** [comment]
Just I mention about the deprecation policy in ESLint core.

- Deprecating a rule happens in a minor version.
- Deprecated rules are frozen. People can continue to use the deprecated rules, but they don't change the deprecated rules even if bugs exist.
- They make renaming a rule in a minor version if it's needed because renaming is a combination of deprecating a rule and adding a new rule. (it's *not* alias.)
- Some precedents exist in core. For example, `no-comma-dangle` had been renamed to `comma-dangle` because it's a rule that enforces the style about trailing commas rather than that disallows trailing commas.

> https://eslint.org/docs/user-guide/rule-deprecation

**Comment by j-f1** [comment]
Should we do a 2.0 and get rid of all the `no-` rules?

**Comment by bradzacher** [comment]
A number of the `no-` rules make sense as things that won't be turned on and have no inverse, so a blanket rule against `no-` prefixes is a bad idea.
I agree though that it makes sense for us to advise against that naming if it's feasible that we will add inverse functionality to a rule.

Going through the list (checkmark = imo should be renamed):
- [ ] `no-angle-bracket-type-assertion`
  - IMO shouldn't have an inverse because the `<cast>value` syntax is essentially deprecated.
- [X] `no-array-constructor`
  - has no inverse
  - maybe should be renamed though because it sounds like you shouldn't use any array constructors, when it's only enforcing correct usage.
- [ ] `no-empty-interface`
  - has no inverse
- [ ] `no-explicit-any`
  - has no inverse
- [X] `no-extraneous-class`
  - I doubt anyone would ever want the inverse of this unless they absolutely love java.
  - would be good to rename this so it's clearer though (no-class-as-namespace?)
- [ ] `no-inferrable-types`
  - could feasibly have an inverse but I don't think it makes sense to have one
- [ ] `no-misused-new`
  - has no inverse
- [ ] `no-namespace`
  - would people want an inverse for this? I haven't ever seen namespace used outside of definition files.
- [X] `no-non-null-assertion`
  - could have an inverse? (would require typechecking though)
- [ ] `no-object-literal-type-assertion`
  - inverse seems weird here due to the type-unsoundness which `as X` introduces?
- [X] `no-parameter-properties`
  - *needs* inverse
- [ ] `no-this-alias`
  - has no inverse
- [ ] `no-triple-slash-reference`
  - has no inverse (because `/// <ref />` is essentially deprecated)
- [X] `no-type-alias`
  - *needs* inverse
- [ ] `no-unnecessary-type-assertion`
  - has no inverse
- [ ] `no-unused-vars`
  - has no inverse
- [ ] `no-use-before-define`
  - has no inverse
- [ ] `no-useless-constructor`
  - has no inverse (well technically it does - force every class to have at least a useless constructor, but that seems like a really poor practice).
- [ ] `no-var-requires`
  - has no inverse that makes sense in TS.

**Comment by armano2** [comment]
i don't see point in rules like:
- no-non-null-assertion: inverted -> what it will check? that you need null assertion?
- no-type-alias -> there should no be inverse, we should have 2nd rule: `no-interfaces` or something like that

it's better to keep rules simple

**Comment by bradzacher** [comment]
the problem with two rules is now you can have a state where you're both not allowed to use interfaces and not allowed to use type aliases.

we can have a rule `type-definition-style` which has an option to pick either `interface` or `type-alias`. It's still simple and prevents invalid states.

**Comment by gavinbarron** [comment]
Regarding the change to add options to `explicit-member-visibility` in #214 I guess the better name for the rule would be `member-visibility` I assume that this would however be a breaking change. 
What's the process/plan for implementing these changes as a whole?

**Comment by JoshuaKGoldberg** [comment]
Marking as accepting breaking change PRs per the list in https://github.com/typescript-eslint/typescript-eslint/issues/203#issuecomment-460360230.

**Comment by nickmccurdy** [comment]
Can we at least keep the `no-` prefix for extension rules? I don't think the benefits of simplifying the APIs apply to rules that are already in ESLint, and this makes migrating from ESLint rules more confusing.
